lib/fs/spifs/spifs.c

```diff
@@ -0,0 +1,1146 @@
+        return ERR_TOO_BIG;
+    }
+
+    // Create a mock spifs_t for the purposes of formatting the fs.
+    spifs_t spifs = {
+        .page_size = page_size,
+        .page_count = page_count,
+        .blocks_per_page = divpow2(page_size, dev->block_shift),
+        .generation = 1,
+        .num_entries = num_toc_entries,
+        .toc_position = FRONT_TOC,
+        .dev = dev,
+    };
+    spifs.page = memalign(CACHE_LINE, page_size);
+    list_initialize(&spifs.files);
+    list_initialize(&spifs.dcookies);
+    mutex_init(&spifs.lock);
+
+    spifs_file_t f_toc;
+    f_toc.metadata.page_idx = 0;
+    f_toc.metadata.length = spifs_args->toc_pages * page_size;
+    f_toc.metadata.capacity = spifs_args->toc_pages * page_size;
+    f_toc.fs_handle = &spifs;
+    memset(f_toc.metadata.filename, 0, MAX_FILENAME_LENGTH);
+    strlcpy(f_toc.metadata.filename, FRONT_TOC_LABEL, MAX_FILENAME_LENGTH);
+
+    spifs_file_t b_toc;
+    b_toc.metadata.page_idx = page_count - spifs_args->toc_pages;
+    b_toc.metadata.length = spifs_args->toc_pages * page_size;
+    b_toc.metadata.capacity = spifs_args->toc_pages * page_size;
+    b_toc.fs_handle = &spifs;
+    memset(b_toc.metadata.filename, 0, MAX_FILENAME_LENGTH);
+    strlcpy(b_toc.metadata.filename, BACK_TOC_LABEL, MAX_FILENAME_LENGTH);
+
+    spifs_add_ascending(&spifs, &f_toc);
+    spifs_add_ascending(&spifs, &b_toc);
+
+    // Commit the first toc.
+    err = spifs_commit_toc(&spifs);
+    if (err != NO_ERROR)
+        goto err;
+
+    // Commit the other toc.
+    err = spifs_commit_toc(&spifs);
+    if (err != NO_ERROR)
+        goto err;
+
+err:
+    free(spifs.page);
+
+    return err;
+}
+
+static status_t spifs_mount(bdev_t *dev, fscookie **cookie)
+{
+    status_t status;
+
+    LTRACEF("dev %p, cookie %p\n", dev, cookie);
+
+    spifs_t *spifs = malloc(sizeof(*spifs));
+    if (!spifs) {
+        return ERR_NO_MEMORY;
+    }
+
+    status = get_device_page_info(dev, &spifs->page_size, &spifs->page_count);
+    if (status != NO_ERROR) {
+        free(spifs);
+        return status;
+    }
+
+    spifs->blocks_per_page = divpow2(spifs->page_size, dev->block_shift);
+
+    spifs->page = memalign(CACHE_LINE, spifs->page_size);
+    if (!spifs->page) {
+        free(spifs);
+        return ERR_NO_MEMORY;
+    }
+
+    spifs->dev = dev;
+
+    list_initialize(&spifs->files);
+    list_initialize(&spifs->dcookies);
+    mutex_init(&spifs->lock);
+
+    // Determine which of the two Table of Contents we should use.
+    uint32_t f_toc_generation = get_toc_generation(spifs, FRONT_TOC);
+    uint32_t b_toc_generation = get_toc_generation(spifs, BACK_TOC);
+
+    if (f_toc_generation == CORRUPT_TOC && b_toc_generation == CORRUPT_TOC) {
+        // Both ToCs are corrupt.
+        status = ERR_CRC_FAIL;
+        goto err;
+    }
+
+    spifs->toc_position =
+        f_toc_generation > b_toc_generation ? FRONT_TOC : BACK_TOC;
+    spifs->generation = MAX(f_toc_generation, b_toc_generation);
```

Comment:
not that'd it'd be a problem but you can also handle the wraparound case if you assume the generations are within 31 bits of each other. I think you basically subtract them from each other, and based on the difference you can tell which one is higher, even if it wraps.